src/jit/codegenxarch.cpp

```diff
@@ -7549,6 +7556,232 @@ void CodeGen::genSetRegToCond(regNumber dstReg, GenTreePtr tree)
     }
 }
 
+
+void CodeGen::genOverflowCheck_I8toI4(regNumber hiReg, regNumber loReg)
+{
+
+    // conv.ovf.i8.i4
+    // Generate the following sequence
+    // test loDWord, loDWord   // set flags
+    // jl neg
+    // pos: test hiDWord, hiDWord   // set flags
+    // jne ovf
+    // jmp done
+    // neg: cmp hiDWord, 0xFFFFFFFF
+    // jne ovf
+    // done:
+
+    emitJumpKind jmpNotEqual = genJumpKindForOper(GT_NE, CK_SIGNED);
+    emitJumpKind jmpLTS = genJumpKindForOper(GT_LT, CK_SIGNED);
+
+    instGen_Compare_Reg_To_Zero(EA_4BYTE, loReg);
+
+    BasicBlock * neg;
+    BasicBlock * done;
+
+    neg = genCreateTempLabel();
+    done = genCreateTempLabel();
+
+    // Is the loDWord positive or negative
+    inst_JMP(jmpLTS, neg);
+
+    // If loDWord is positive, hiDWord should be 0 (sign extended loDWord)
+    instGen_Compare_Reg_To_Zero(EA_4BYTE, hiReg);
+
+    genJumpToThrowHlpBlk(jmpNotEqual, SCK_OVERFLOW);
+    inst_JMP(EJ_jmp, done);
+
+    // If loDWord is negative, hiDWord should be -1 (sign extended loDWord)
+
+    genDefineTempLabel(neg);
+
+    inst_RV_IV(INS_cmp, hiReg, 0xFFFFFFFFL, EA_4BYTE);
+
+    genJumpToThrowHlpBlk(jmpNotEqual, SCK_OVERFLOW);
+
+    // Done
+
+    genDefineTempLabel(done);
+}
+
+void CodeGen::genOverflowCheck_I8toU4(regNumber hiReg, regNumber loReg)
+{
+    // Just check that the upper DWord is 0
+
+    instGen_Compare_Reg_To_Zero(EA_4BYTE, hiReg); // set flags
+
+    emitJumpKind jmpNotEqual = genJumpKindForOper(GT_NE, CK_SIGNED);
+    genJumpToThrowHlpBlk(jmpNotEqual, SCK_OVERFLOW);
+}
+
+void CodeGen::genOverflowCheck_U8toI4(regNumber hiReg, regNumber loReg)
+{
+    // ; if low_word < 0 raise overflow
+    // 02D02DE3  mov         eax, dword ptr ds : [0169C8E8h]
+    // 02D02DE8  test        eax, eax
+    // 02D02DEA  jl          02D02DF9
+    // ; if high_word != 0 raise overflow
+    // 02D02DEC  cmp         dword ptr ds : [169C8ECh], 0
+    // 02D02DF3  jne         02D02DF9
+    // 02D02DF5  pop         ebp
+    // 02D02DF6  ret         8
+    // ; overflow exception
+    // 02D02DF9  call        73BECA1D 
+    emitJumpKind jmpNotEqual = genJumpKindForOper(GT_NE, CK_SIGNED);
+    emitJumpKind jmpLTS = genJumpKindForOper(GT_LT, CK_SIGNED);
+
+    instGen_Compare_Reg_To_Zero(EA_4BYTE, loReg);
+    genJumpToThrowHlpBlk(jmpLTS, SCK_OVERFLOW);
+
+    instGen_Compare_Reg_To_Zero(EA_4BYTE, hiReg);
+    genJumpToThrowHlpBlk(jmpNotEqual, SCK_OVERFLOW);
+}
+
+void CodeGen::genLongToIntCast(GenTreePtr treeNode)
+{
+    assert(treeNode->OperGet() == GT_CAST);
+    assert(treeNode->gtOverflow());
+
+    GenTreePtr  castOp = treeNode->gtCast.CastOp();
+    bool        checkOverflow = treeNode->gtOverflow();
+    var_types   dstType = treeNode->CastToType();
+    bool        isUnsignedDst = varTypeIsUnsigned(dstType);
+    var_types   srcType = genActualType(castOp->TypeGet());
+    bool        isUnsignedSrc = varTypeIsUnsigned(srcType);
```

Comment:
See the `GTF_UNSIGNED` comment - the flags applies to the source type:
```
#define GTF_UNSIGNED 0x00100000 // with GT_CAST:   the source operand is an unsigned type
                                // with operators: the specified node is an unsigned operator
```
IMO it would have been clearer if the comment didn't special case `GT_CAST`. Ultimately the source type signedness is irrelevant. If `GTF_UNSIGNED` is present the cast zero extends the source, otherwise it sign extends the source.

The destination type can be unsigned. Normally the sign of the destination type is irrelevant as well but in the case of `GT_CAST` we actually have 2 casts in one:
* one from source type to destination type (affected by `GTF_UNSIGNED`)
* one from destination type to the cast node type - `TYP_INT` usually (affected by the destination type signedness)